crates/syntax/src/validation.rs

```diff
@@ -339,35 +337,73 @@ fn validate_trait_object_fn_ptr_ret_ty(ty: ast::FnPtrType, errors: &mut Vec<Synt
     }
 }
 
-fn validate_trait_object_ty(ty: ast::DynTraitType) -> Option<SyntaxError> {
-    let tbl = ty.type_bound_list()?;
-    let bounds_count = tbl.bounds().count();
+fn validate_object_ty(
+    bounds: impl Iterator<Item = ast::TypeBound>,
+    preceding_token: Option<SyntaxToken>,
+    error_no_trait: &str,
+    ty_syntax_range: TextRange,
+    check_for_and_token: bool,
+) -> Option<SyntaxError> {
+    let bounds: Vec<_> = bounds.collect();
+    let bounds_count = bounds.len();
+
+    if bounds.is_empty()
+        || bounds.iter().all(|b| matches!(b.kind(), ast::TypeBoundKind::Lifetime(_)))
+    {
+        return Some(SyntaxError::new(error_no_trait, ty_syntax_range));
+    }
```

Comment:
we definitely don't want to allocate here so
```suggestion
    if !bounds.iter().any(|b| !matches!(b.kind(), ast::TypeBoundKind::Lifetime(_)))
    {
        return Some(SyntaxError::new(error_no_trait, ty_syntax_range));
    }
    let bounds_count = bounds.len();
```